hs-abci-sdk/src/Tendermint/SDK/BaseApp/Metrics/Prometheus.hs

```diff
@@ -0,0 +1,187 @@
+                                                                Sem, interpretH,
+                                                                pureT, raise,
+                                                                runT)
+import           Polysemy.Reader                               (Reader (..),
+                                                                ask)
+import           System.Environment                            (lookupEnv)
+import qualified System.Metrics.Prometheus.Concurrent.Registry as Registry
+import qualified System.Metrics.Prometheus.Http.Scrape         as Http
+import qualified System.Metrics.Prometheus.Metric.Counter      as Counter
+import qualified System.Metrics.Prometheus.Metric.Histogram    as Histogram
+import qualified System.Metrics.Prometheus.MetricId            as MetricId
+import           Tendermint.SDK.BaseApp.Metrics                (CountName (..), HistogramName (..),
+                                                                Metrics (..),
+                                                                observeHistogram)
+import qualified Text.Read                                     as T
+
+--------------------------------------------------------------------------------
+-- eval
+--------------------------------------------------------------------------------
+
+evalMetrics
+  :: Member (Embed IO) r
+  => MetricsState
+  -> Sem (Metrics ': r) a
+  -> Sem r a
+evalMetrics state@MetricsState{..} = do
+  interpretH (\case
+    IncCount ctrName -> do
+      let c@MetricIdentifier{..} = countToIdentifier ctrName
+          cid = metricIdStorable c
+          cMetricIdName = MetricId.Name metricIdName
+      liftIO $ modifyMVar_ metricsCounters $ \counterMap ->
+        case Map.lookup cid counterMap of
+          Nothing -> do
+            newCtr <- liftIO $
+              Registry.registerCounter cMetricIdName metricIdLabels metricsRegistry
+            pure $ insert cid newCtr counterMap
+          Just ctr -> do
+            liftIO $ Counter.inc ctr
+            pure counterMap
+      pureT ()
+
+    ObserveHistogram histName val -> do
+      let h@MetricIdentifier{..} = histogramToIdentifier histName
+          hid = metricIdStorable h
+          hMetricIdName = MetricId.Name metricIdName
+      liftIO $ modifyMVar_ metricsHistograms $ \histMap ->
+        case Map.lookup hid histMap of
+          Nothing -> do
+            newHist <- liftIO $
+              Registry.registerHistogram hMetricIdName metricIdLabels metricIdHistoBuckets metricsRegistry
+            pure $ insert hid newHist histMap
+          Just hist -> do
+            liftIO $ Histogram.observe val hist
+            pure histMap
+      pureT ()
+
+    WithTimer histName action -> do
+      start <- liftIO $ getCurrentTime
+      a <- runT action
+      end <- liftIO $ getCurrentTime
+      let time = fromRational . (* 1000.0) . toRational $ (end `diffUTCTime` start)
+      evalMetrics state (observeHistogram histName time)
```

Comment:
To me this is further evidence that observeHistogram should just be a function used in the interpreter written in terms of `IO`, or something else outside the `Metrics` DSL

Reply:
Yeah that makes sense.